src/subtype.c

```diff
@@ -796,90 +818,186 @@ static int check_vararg_length(jl_value_t *v, ssize_t n, jl_stenv_t *e)
-        }
-        if (vx && !vy) {
-            if (!check_vararg_length(xi, ly+1-lx, e))
-                return 0;
-            jl_tvar_t *p1=NULL, *p2=NULL;
-            xi = unwrap_2_unionall(xi, &p1, &p2);
-            jl_value_t *N = jl_tparam1(xi);
-            if (N == (jl_value_t*)p1 || N == (jl_value_t*)p2)
-                return 0;
-            if (j >= ly) return 1;
-            xi = jl_tparam0(xi);
+
+    if (!jl_is_datatype(env->vtx)) {
+        jl_value_t *yl = yp1;
+        if (jl_is_typevar(yl)) {
+            jl_varbinding_t *ylv = lookup(e, (jl_tvar_t*)yl);
+            if (ylv)
+                yl = ylv->lb;
         }
-        else if (j >= ly) {
+        if (jl_is_long(yl)) {
+            // The length of the x tuple is unconstrained, but the
+            // length of the y tuple is now fixed (this could have happened
+            // as a result of the subtype call above).
             return 0;
         }
-        if (!vx && vy) {
-            jl_tvar_t *p1=NULL, *p2=NULL;
-            yi = jl_tparam0(unwrap_2_unionall(yi, &p1, &p2));
-            if (yi == (jl_value_t*)p1 || yi == (jl_value_t*)p2)
-                yi = ((jl_tvar_t*)yi)->ub;
-            if (!vvx && yi == (jl_value_t*)jl_any_type)
-                break;  // if y ends in `Vararg{Any}` skip checking everything
-        }
-        if (vx && vy) {
-            jl_tvar_t *yp1=NULL, *yp2=NULL;
-            jl_value_t *yva = unwrap_2_unionall(yi, &yp1, &yp2);
-            jl_tvar_t *xp1=NULL, *xp2=NULL;
-            jl_value_t *xva = unwrap_2_unionall(xi, &xp1, &xp2);
-            if ((jl_value_t*)xp1 == jl_tparam1(xva) || (jl_value_t*)xp2 == jl_tparam1(xva)) {
-                // check for unconstrained vararg on left, constrained on right
-                if (jl_is_long(jl_tparam1(yva)))
-                    return 0;
-                if (jl_is_typevar(jl_tparam1(yva))) {
-                    jl_varbinding_t *ylv = lookup(e, (jl_tvar_t*)jl_tparam1(yva));
-                    if (ylv && jl_is_long(ylv->lb))
-                        return 0;
-                }
-            }
+    }
 
-            // skip testing element type if vararg lengths are 0
-            if (jl_is_datatype(xi)) {
-                jl_value_t *xl = jl_tparam1(xi);
-                if (jl_is_typevar(xl)) {
-                    jl_varbinding_t *xlv = lookup(e, (jl_tvar_t*)xl);
-                    if (xlv && jl_is_long(xlv->lb) && jl_unbox_long(xlv->lb) == 0)
-                        break;
+    // Vararg{T,N} <: Vararg{T2,N2}; equate N and N2
+    e->invdepth++;
+    if (jl_is_long(xp1) && env->vx != 1)
+        xp1 = jl_box_long(jl_unbox_long(xp1) - env->vx + 1);
+    if (jl_is_long(yp1) && env->vy != 1)
+        yp1 = jl_box_long(jl_unbox_long(yp1) - env->vy + 1);
+    int ans = forall_exists_equal(xp1, yp1, e);
+    e->invdepth--;
+    return ans;
+}
+
+static int subtype_tuple_tail(struct subtype_tuple_env *env, int8_t R, jl_stenv_t *e, int param)
+{
+loop:
+    // while (i <= lx) {
```

Comment:
maybe put this comment on the previous line, so be clearer that it's a comment and not deadcode?